This patch is based on a revert of upstream commits
6e11c77384989726bb4f412a0e23b59c27222c34 and
ae91d4ed41da98bdfa16041dbc6cd30287920120, adapted for use in the Debian package
by Peter Michael Green.

Index: home/src/lib.rs
===================================================================
--- home.orig/src/lib.rs
+++ home/src/lib.rs
@@ -44,11 +44,11 @@ use std::path::{Path, PathBuf};
 ///
 /// Returns the value of the `USERPROFILE` environment variable if it is set
 /// **and** it is not an empty string. Otherwise, it tries to determine the
-/// home directory by invoking the [`SHGetKnownFolderPath`][shgkfp] function with
-/// [`FOLDERID_Profile`][knownfolderid].
+/// home directory by invoking the [`SHGetFolderPathW`][shgfp] function with
+/// [`CSIDL_PROFILE`][csidl].
 ///
-/// [shgkfp]: https://learn.microsoft.com/en-us/windows/win32/api/shlobj_core/nf-shlobj_core-shgetknownfolderpath
-/// [knownfolderid]: https://learn.microsoft.com/en-us/windows/win32/shell/knownfolderid
+/// [shgfp]: https://docs.microsoft.com/en-us/windows/win32/api/shlobj_core/nf-shlobj_core-shgetfolderpathw
+/// [csidl]: https://learn.microsoft.com/en-us/windows/win32/shell/csidl
 ///
 /// # Examples
 ///
Index: home/src/windows.rs
===================================================================
--- home.orig/src/windows.rs
+++ home/src/windows.rs
@@ -2,12 +2,10 @@ use std::env;
 use std::ffi::OsString;
 use std::os::windows::ffi::OsStringExt;
 use std::path::PathBuf;
-use std::ptr;
-use std::slice;
 
-use windows_sys::Win32::Foundation::S_OK;
-use windows_sys::Win32::System::Com::CoTaskMemFree;
-use windows_sys::Win32::UI::Shell::{FOLDERID_Profile, SHGetKnownFolderPath, KF_FLAG_DONT_VERIFY};
+use winapi::shared::minwindef::MAX_PATH;
+use winapi::shared::winerror::S_OK;
+use winapi::um::shlobj::{SHGetFolderPathW, CSIDL_PROFILE};
 
 pub fn home_dir_inner() -> Option<PathBuf> {
     env::var_os("USERPROFILE")
@@ -19,19 +17,15 @@ pub fn home_dir_inner() -> Option<PathBu
 #[cfg(not(target_vendor = "uwp"))]
 fn home_dir_crt() -> Option<PathBuf> {
     unsafe {
-        let mut path = ptr::null_mut();
-        match SHGetKnownFolderPath(&FOLDERID_Profile, KF_FLAG_DONT_VERIFY as u32, 0, &mut path) {
+        let mut path: Vec<u16> = Vec::with_capacity(MAX_PATH as usize);
+        match SHGetFolderPathW(0, CSIDL_PROFILE as i32, 0, 0, path.as_mut_ptr()) {
             S_OK => {
-                let path_slice = slice::from_raw_parts(path, wcslen(path));
-                let s = OsString::from_wide(&path_slice);
-                CoTaskMemFree(path.cast());
+                let len = wcslen(path.as_ptr());
+                path.set_len(len);
+                let s = OsString::from_wide(&path);
                 Some(PathBuf::from(s))
             }
-            _ => {
-                // Free any allocated memory even on failure. A null ptr is a no-op for `CoTaskMemFree`.
-                CoTaskMemFree(path.cast());
-                None
-            }
+            _ => None,
         }
     }
 }
Index: home/Cargo.toml
===================================================================
--- home.orig/Cargo.toml
+++ home/Cargo.toml
@@ -28,12 +28,12 @@ readme = "README.md"
 license = "MIT OR Apache-2.0"
 repository = "https://github.com/rust-lang/cargo"
 
-[target."cfg(windows)".dependencies.windows-sys]
-version = "0.52"
+[target."cfg(windows)".dependencies.winapi]
+version = "0.3"
 features = [
-    "Win32_Foundation",
-    "Win32_UI_Shell",
-    "Win32_System_Com",
+    "shlobj",
+    "std",
+    "winerror",
 ]
 
 [lints.clippy]
